debug/resolver/interceptorResolver.go

```diff
@@ -0,0 +1,347 @@
+package resolver
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/debug"
+	"github.com/ElrondNetwork/elrond-go/storage/lrucache"
+)
+
+const requestEvent = "request"
+const resolveEvent = "resolve"
+const minIntervalInSeconds = 1
+const minThresholdResolve = 1
+const minThresholdRequests = 1
+const newLineChar = "\n"
+
+var log = logger.GetOrCreate("debug/resolver")
+
+type event struct {
+	eventType    string
+	hash         []byte
+	topic        string
+	numReqIntra  int
+	numReqCross  int
+	numReceived  int
+	numProcessed int
+	lastErr      error
+	numPrints    int
+}
+
+func (ev *event) String() string {
+	strErr := ""
+	if ev.lastErr != nil {
+		strErr = ev.lastErr.Error()
+	}
+	return fmt.Sprintf("type: %s, topic: %s, hash: %s, numReqIntra: %d, numReqCross: %d, "+
+		"numReceived: %d, numProcessed: %d, last err: %s",
+		ev.eventType,
+		ev.topic,
+		logger.DisplayByteSlice(ev.hash),
+		ev.numReqIntra,
+		ev.numReqCross,
+		ev.numReceived,
+		ev.numProcessed,
+		strErr,
+	)
+}
+
+type interceptorResolver struct {
+	mutCriticalArea      sync.RWMutex
+	cache                *lrucache.LRUCache
+	intervalAutoPrint    time.Duration
+	requestsThreshold    int
+	resolveFailThreshold int
+	maxNumPrints         int
+	printEventHandler    func(data string)
+}
+
+// NewInterceptorResolver creates a new interceptorResolver able to hold requested-intercepted information
+func NewInterceptorResolver(config config.InterceptorResolverDebugConfig) (*interceptorResolver, error) {
+	cache, err := lrucache.NewCache(config.CacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("%w when creating NewInterceptorResolver", err)
+	}
+
+	ir := &interceptorResolver{
+		cache: cache,
+	}
+
+	err = ir.parseConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	ir.printEventHandler = ir.printEvent
+	if config.EnableAutoPrint {
+		go ir.autoPrint()
+	}
+
+	return ir, nil
+}
+
+func (ir *interceptorResolver) parseConfig(config config.InterceptorResolverDebugConfig) error {
+	if !config.EnableAutoPrint {
+		return nil
+	}
+	if config.IntervalAutoPrintInSeconds < minIntervalInSeconds {
+		return fmt.Errorf("%w for IntervalAutoPrintInSeconds, minimum is %d", debug.ErrInvalidValue, minIntervalInSeconds)
+	}
+	if config.NumRequestsThreshold < minThresholdRequests {
+		return fmt.Errorf("%w for NumRequestsThreshold, minimum is %d", debug.ErrInvalidValue, minThresholdRequests)
+	}
+	if config.NumResolveFailureThreshold < minThresholdResolve {
+		return fmt.Errorf("%w for NumResolveFailureThreshold, minimum is %d", debug.ErrInvalidValue, minThresholdResolve)
+	}
+	if config.DebugLineExpiration <= 0 {
```

Comment:
Missmatch between check (<=0) and comment (minimum is 1)

Reply:
yeah, but is correct as the error would have been yield for negative values or 0. For values  equal or greater than 1 it would have been fine. Anyway, added const, refactored.